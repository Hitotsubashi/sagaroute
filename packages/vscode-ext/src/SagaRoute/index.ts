import BundledSagaRoute, { RoutingOption } from '@sagaroute/react';
import * as vscode from 'vscode';
import mergeOption from '@sagaroute/react/lib/utils/mergeOption';
import loggingHooks from './hooks/logging';
import writingHooks from './hooks/writing';
import cloneDeep from 'lodash/cloneDeep';
import statisticPathHooks from './hooks/statistic-path';
import getWarningManager from '../WarningManager';
import getLogging from '../Logging';

const workspaceRootFolderPath = vscode.workspace.workspaceFolders![0].uri.fsPath;

const mergedHooks = [loggingHooks, writingHooks, statisticPathHooks].reduce(
  (pre, cur) => mergeOption(pre, { hooks: cur }),
  {} as RoutingOption,
).hooks;

const sagaRouteConfig: RoutingOption = {
  rootPath: workspaceRootFolderPath,
  hooks: mergedHooks,
  onWarning(message) {
    const warningManager = getWarningManager();
    warningManager.addMessage(message);
  },
};

function getSagaRouteConfig() {
  return cloneDeep(sagaRouteConfig);
}

class Sagaroute extends BundledSagaRoute {
  error?: boolean;
}

let sagaRoute: Sagaroute;
// const dependencyName = '@sagaroute/react';

function getSagarouteClass(): typeof Sagaroute {
  return BundledSagaRoute;
  // const logging = getLogging();
  // try {
  //   if (!vscode.workspace.isTrusted) {
  //     logging.logMessage('This workspace is not trusted. Using the bundled sagaroute.');
  //     return BundledSagaRoute;
  //   }
  //   const stats = fs.statSync(path.join(workspaceRootFolderPath, 'package.json'));
  //   if (stats?.isFile()) {
  //     const packageJSON = require(path.join(workspaceRootFolderPath, 'package.json'));
  //     const version =
  //       packageJSON.devDependencies?.[dependencyName] || packageJSON.dependencies?.[dependencyName];
  //     if (version) {
  //       const localSagaroutePath = path.join(
  //         workspaceRootFolderPath,
  //         'node_modules',
  //         dependencyName,
  //       );
  //       if (fs.statSync(localSagaroutePath)) {
  //         logging.logMessage(`Using the local sagaroute:${version}.`);
  //         return require(localSagaroutePath).default;
  //       } else {
  //         logging.logMessage(
  //           `The file [package.json] has "${dependencyName}" in dependency but not found in node_modules. Using the bundled sagaroute.`,
  //         );
  //         return BundledSagaRoute;
  //       }
  //     } else {
  //       logging.logMessage(
  //         `The file [package.json] doesn't has "${dependencyName}" in either dependency or devDependency. Using the bundled sagaroute.`,
  //       );
  //       return BundledSagaRoute;
  //     }
  //   } else {
  //     logging.logMessage('The file [package.json] is not exist. Using the bundled sagaroute.');
  //     return BundledSagaRoute;
  //   }
  // } catch (err) {
  //   logging.logMessage(`ERROR: ${err}. Using the bundled sagaroute.`, 'ERROR');
  //   return BundledSagaRoute;
  // }
}

export default function getSagaRoute() {
  const logging = getLogging();
  if (!sagaRoute) {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    const SagaRoute = getSagarouteClass();
    try {
      sagaRoute = new SagaRoute(getSagaRouteConfig());
    } catch (error: any) {
      logging.logMessage(`Init Sagaroute Error: ${error.message}.`, 'ERROR');
      vscode.window.showErrorMessage(error.message);
      sagaRoute.error = true;
    }
  }
  return sagaRoute;
}

export function rebuildSagaroute() {
  const logging = getLogging();
  try {
    sagaRoute.generateOptions(getSagaRouteConfig());
    sagaRoute.error = false;
  } catch (error: any) {
    logging.logMessage(`Init Sagaroute Error: ${error.message}.`, 'ERROR');
    vscode.window.showErrorMessage(error.message);
    sagaRoute.error = true;
  }
  return sagaRoute;
}
