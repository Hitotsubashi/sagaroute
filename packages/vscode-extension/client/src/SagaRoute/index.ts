import BundledSagaRoute from '@sagaroute/react';
import SagaRoute, { RoutingOption } from '@sagaroute/react';
import * as vscode from 'vscode';
import mergeOption from '@sagaroute/react/lib/utils/mergeOption';
import loggingHooks from './hooks/logging';
import writingHooks from './hooks/writing';
import cloneDeep from 'lodash/cloneDeep';
import statisticPathHooks from './hooks/statistic-path';
import getWarningManager from '../WarningManager';
import getLogging from '../Logging';
import fs from 'fs';
import path from 'path';

const workspaceRootFolderPath = vscode.workspace.workspaceFolders![0].uri.fsPath;

const mergedHooks = [loggingHooks, writingHooks, statisticPathHooks].reduce(
  (pre, cur) => mergeOption(pre, { hooks: cur }),
  {} as RoutingOption,
).hooks;

const sagaRouteConfig: RoutingOption = {
  rootPath: workspaceRootFolderPath,
  hooks: mergedHooks,
  onWarning(message) {
    const warningManager = getWarningManager();
    warningManager.addMessage(message);
  },
};

export function getSagaRouteConfig() {
  return cloneDeep(sagaRouteConfig);
}

let sagaRoute: SagaRoute;
const dependencyName = '@sagaroute/react';

function getSagarouteClass(): typeof BundledSagaRoute {
  const logging = getLogging();
  try {
    if (!vscode.workspace.isTrusted) {
      logging.logMessage('This workspace is not trusted. Using the bundled sagaroute.');
      return BundledSagaRoute;
    }
    const stats = fs.statSync(path.join(workspaceRootFolderPath, 'package.json'));
    if (stats?.isFile()) {
      const packageJSON = require(path.join(workspaceRootFolderPath, 'package.json'));
      const version =
        packageJSON.devDependencies?.[dependencyName] || packageJSON.dependencies?.[dependencyName];
      if (version) {
        const localSagaroutePath = path.join(
          workspaceRootFolderPath,
          'node_modules',
          dependencyName,
        );
        if (fs.statSync(localSagaroutePath)) {
          logging.logMessage(`Using the local sagaroute:${version}.`);
          return require(localSagaroutePath).default;
        } else {
          logging.logMessage(
            `The file [package.json] has "${dependencyName}" in dependency but not found in node_modules. Using the bundled sagaroute.`,
          );
          return BundledSagaRoute;
        }
      } else {
        logging.logMessage(
          `The file [package.json] doesn't has "${dependencyName}" in either dependency or devDependency. Using the bundled sagaroute.`,
        );
        return BundledSagaRoute;
      }
    } else {
      logging.logMessage('The file [package.json] is not exist. Using the bundled sagaroute.');
      return BundledSagaRoute;
    }
  } catch (err) {
    logging.logMessage(`ERROR: ${err}. Using the bundled sagaroute.`, 'ERROR');
    return BundledSagaRoute;
  }
}

export default function getSagaRoute() {
  if (!sagaRoute) {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    const SagaRoute = getSagarouteClass();
    sagaRoute = new SagaRoute(getSagaRouteConfig());
  }
  return sagaRoute;
}
