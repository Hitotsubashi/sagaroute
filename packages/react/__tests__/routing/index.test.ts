import fs from 'fs';
import path from 'path';
import SagaRoute from '@/index';

test('test routing in normal', () => {
  const buildHooksBefore = jest.fn();
  const buildHooksAfter = jest.fn();
  const gatherHooksBefore = jest.fn();
  const gatherHooksBeforeEach = jest.fn();
  const gatherHooksAfterEach = jest.fn();
  const gatherHooksAfter = jest.fn();
  const weaveHooksBefore = jest.fn();
  const weaveHooksBeforeEach = jest.fn();
  const weaveHooksAfterEach = jest.fn();
  const weaveHooksAfter = jest.fn();
  const printHooksParseBefore = jest.fn();
  const printHooksParseAfter = jest.fn();
  const printHooksInjectBefore = jest.fn();
  const printHooksInjectAfter = jest.fn();
  const printHooksWriteBefore = jest.fn();
  const printHooksWriteAfter = jest.fn();
  const outputFile = path.join(__dirname, 'test-fixtures', 'normal-src', 'route.tsx');
  const option = {
    dirpath: path.join(__dirname, 'test-fixtures', 'normal-src/pages'),
    layoutDirPath: path.join(__dirname, 'test-fixtures', 'normal-src/layouts'),
    routeFilePath: outputFile,
    pathRewrite: {
      '^./': '@/',
      '.tsx$': '',
    },
    hooks: {
      build: {
        before: buildHooksBefore,
        after: buildHooksAfter,
      },
      gather: {
        before: gatherHooksBefore,
        beforeEach: gatherHooksBeforeEach,
        afterEach: gatherHooksAfterEach,
        after: gatherHooksAfter,
      },
      weave: {
        before: weaveHooksBefore,
        beforeEach: weaveHooksBeforeEach,
        afterEach: weaveHooksAfterEach,
        after: weaveHooksAfter,
      },
      print: {
        parse: {
          before: printHooksParseBefore,
          after: printHooksParseAfter,
        },
        inject: {
          before: printHooksInjectBefore,
          after: printHooksInjectAfter,
        },
        write: {
          before: printHooksWriteBefore,
          after: printHooksWriteAfter,
        },
      },
    },
  };
  const sagaRoute = new SagaRoute(option);
  sagaRoute.routing();
  const context = fs.readFileSync(outputFile, 'utf-8');
  expect(context).toMatchSnapshot();
  expect(buildHooksBefore.mock.calls).toHaveLength(1);
  expect(buildHooksBefore.mock.calls[0]).toStrictEqual([option]);
  expect(buildHooksAfter.mock.calls).toHaveLength(1);
  expect(buildHooksAfter.mock.calls[0]).toStrictEqual([
    {
      ...option,
      pathRewrite: [
        [new RegExp('^./'), '@/'],
        [new RegExp('.tsx$'), ''],
      ],
      relativeDirpath: path.join('..', 'pages'),
      relativeLayoutDirPath: path.join('..', 'layouts'),
      hooks: {
        build: {
          before: [buildHooksBefore],
          after: [buildHooksAfter],
        },
        gather: {
          before: [gatherHooksBefore],
          beforeEach: [gatherHooksBeforeEach],
          afterEach: [gatherHooksAfterEach],
          after: [gatherHooksAfter],
        },
        weave: {
          before: [weaveHooksBefore],
          beforeEach: [weaveHooksBeforeEach],
          afterEach: [weaveHooksAfterEach],
          after: [weaveHooksAfter],
        },
        print: {
          parse: {
            before: [printHooksParseBefore],
            after: [printHooksParseAfter],
          },
          inject: {
            before: [printHooksInjectBefore],
            after: [printHooksInjectAfter],
          },
          write: {
            before: [printHooksWriteBefore],
            after: [printHooksWriteAfter],
          },
        },
      },
    },
  ]);

  expect(gatherHooksBefore.mock.calls).toHaveLength(1);
  expect(gatherHooksBeforeEach.mock.calls).toHaveLength(3);
  expect(gatherHooksAfterEach.mock.calls).toHaveLength(3);
  expect(gatherHooksAfter.mock.calls).toHaveLength(1);

  expect(weaveHooksBefore.mock.calls).toHaveLength(1);
  expect(weaveHooksBeforeEach.mock.calls).toHaveLength(3);
  expect(weaveHooksAfterEach.mock.calls).toHaveLength(3);
  expect(weaveHooksAfter.mock.calls).toHaveLength(1);

  expect(printHooksParseBefore.mock.calls).toHaveLength(1);
  expect(printHooksParseAfter.mock.calls).toHaveLength(1);
  expect(printHooksInjectBefore.mock.calls).toHaveLength(1);
  expect(printHooksInjectAfter.mock.calls).toHaveLength(1);
  expect(printHooksWriteBefore.mock.calls).toHaveLength(1);
  expect(printHooksWriteAfter.mock.calls).toHaveLength(1);
});

test('test skip gather,weave and print', () => {
  const buildHooksBefore = jest.fn();
  const buildHooksAfter = jest.fn();
  const gatherHooksBefore = jest.fn(() => null);
  const gatherHooksAfter = jest.fn();
  const weaveHooksBefore = jest.fn();
  const weaveHooksAfter = jest.fn();
  const printHooksParseBefore = jest.fn();
  const printHooksParseAfter = jest.fn();
  const printHooksInjectBefore = jest.fn();
  const printHooksInjectAfter = jest.fn();
  const printHooksWriteBefore = jest.fn();
  const printHooksWriteAfter = jest.fn();
  const outputFile = path.join(__dirname, 'test-fixtures', 'normal-src', 'route.tsx');
  const option = {
    dirpath: path.join(__dirname, 'test-fixtures', 'normal-src/pages'),
    layoutDirPath: path.join(__dirname, 'test-fixtures', 'normal-src/layouts'),
    routeFilePath: outputFile,
    pathRewrite: {
      '^./': '@/',
    },
    hooks: {
      build: {
        before: buildHooksBefore,
        after: buildHooksAfter,
      },
      gather: {
        before: gatherHooksBefore,
        after: gatherHooksAfter,
      },
      weave: {
        before: weaveHooksBefore,
        after: weaveHooksAfter,
      },
      print: {
        parse: {
          before: printHooksParseBefore,
          after: printHooksParseAfter,
        },
        inject: {
          before: printHooksInjectBefore,
          after: printHooksInjectAfter,
        },
        write: {
          before: printHooksWriteBefore,
          after: printHooksWriteAfter,
        },
      },
    },
  };
  const sagaRoute = new SagaRoute(option);
  sagaRoute.routing();
  expect(buildHooksBefore.mock.calls).toHaveLength(1);
  expect(buildHooksAfter.mock.calls).toHaveLength(1);

  expect(gatherHooksBefore.mock.calls).toHaveLength(1);
  expect(gatherHooksAfter.mock.calls).toHaveLength(0);

  expect(weaveHooksBefore.mock.calls).toHaveLength(0);
  expect(weaveHooksAfter.mock.calls).toHaveLength(0);

  expect(printHooksParseBefore.mock.calls).toHaveLength(0);
  expect(printHooksParseAfter.mock.calls).toHaveLength(0);
  expect(printHooksInjectBefore.mock.calls).toHaveLength(0);
  expect(printHooksInjectAfter.mock.calls).toHaveLength(0);
  expect(printHooksWriteBefore.mock.calls).toHaveLength(0);
  expect(printHooksWriteAfter.mock.calls).toHaveLength(0);
});

test('test skip weave and print', () => {
  const buildHooksBefore = jest.fn();
  const buildHooksAfter = jest.fn();
  const gatherHooksBefore = jest.fn();
  const gatherHooksAfter = jest.fn();
  const weaveHooksBefore = jest.fn(() => null);
  const weaveHooksAfter = jest.fn();
  const printHooksParseBefore = jest.fn();
  const printHooksParseAfter = jest.fn();
  const printHooksInjectBefore = jest.fn();
  const printHooksInjectAfter = jest.fn();
  const printHooksWriteBefore = jest.fn();
  const printHooksWriteAfter = jest.fn();
  const outputFile = path.join(__dirname, 'test-fixtures', 'normal-src', 'route.tsx');
  const option = {
    dirpath: path.join(__dirname, 'test-fixtures', 'normal-src/pages'),
    layoutDirPath: path.join(__dirname, 'test-fixtures', 'normal-src/layouts'),
    routeFilePath: outputFile,
    pathRewrite: {
      '^./': '@/',
    },
    hooks: {
      build: {
        before: buildHooksBefore,
        after: buildHooksAfter,
      },
      gather: {
        before: gatherHooksBefore,
        after: gatherHooksAfter,
      },
      weave: {
        before: weaveHooksBefore,
        after: weaveHooksAfter,
      },
      print: {
        parse: {
          before: printHooksParseBefore,
          after: printHooksParseAfter,
        },
        inject: {
          before: printHooksInjectBefore,
          after: printHooksInjectAfter,
        },
        write: {
          before: printHooksWriteBefore,
          after: printHooksWriteAfter,
        },
      },
    },
  };
  const sagaRoute = new SagaRoute(option);
  sagaRoute.routing();
  expect(buildHooksBefore.mock.calls).toHaveLength(1);
  expect(buildHooksAfter.mock.calls).toHaveLength(1);

  expect(gatherHooksBefore.mock.calls).toHaveLength(1);
  expect(gatherHooksAfter.mock.calls).toHaveLength(1);

  expect(weaveHooksBefore.mock.calls).toHaveLength(1);
  expect(weaveHooksAfter.mock.calls).toHaveLength(0);

  expect(printHooksParseBefore.mock.calls).toHaveLength(0);
  expect(printHooksParseAfter.mock.calls).toHaveLength(0);
  expect(printHooksInjectBefore.mock.calls).toHaveLength(0);
  expect(printHooksInjectAfter.mock.calls).toHaveLength(0);
  expect(printHooksWriteBefore.mock.calls).toHaveLength(0);
  expect(printHooksWriteAfter.mock.calls).toHaveLength(0);
});

test('test skip print.parse', () => {
  const printHooksParseBefore = jest.fn(() => null);
  const printHooksParseAfter = jest.fn();
  const printHooksInjectBefore = jest.fn();
  const printHooksInjectAfter = jest.fn();
  const printHooksWriteBefore = jest.fn();
  const printHooksWriteAfter = jest.fn();
  const outputFile = path.join(__dirname, 'test-fixtures', 'normal-src', 'route.tsx');
  const option = {
    dirpath: path.join(__dirname, 'test-fixtures', 'normal-src/pages'),
    layoutDirPath: path.join(__dirname, 'test-fixtures', 'normal-src/layouts'),
    routeFilePath: outputFile,
    pathRewrite: {
      '^./': '@/',
    },
    hooks: {
      print: {
        parse: {
          before: printHooksParseBefore,
          after: printHooksParseAfter,
        },
        inject: {
          before: printHooksInjectBefore,
          after: printHooksInjectAfter,
        },
        write: {
          before: printHooksWriteBefore,
          after: printHooksWriteAfter,
        },
      },
    },
  };
  const sagaRoute = new SagaRoute(option);
  sagaRoute.routing();
  expect(printHooksParseBefore.mock.calls).toHaveLength(1);
  expect(printHooksParseAfter.mock.calls).toHaveLength(0);
  expect(printHooksInjectBefore.mock.calls).toHaveLength(0);
  expect(printHooksInjectAfter.mock.calls).toHaveLength(0);
  expect(printHooksWriteBefore.mock.calls).toHaveLength(0);
  expect(printHooksWriteAfter.mock.calls).toHaveLength(0);
});

test('test skip print.inject', () => {
  const printHooksParseBefore = jest.fn();
  const printHooksParseAfter = jest.fn();
  const printHooksInjectBefore = jest.fn(() => null);
  const printHooksInjectAfter = jest.fn();
  const printHooksWriteBefore = jest.fn();
  const printHooksWriteAfter = jest.fn();
  const outputFile = path.join(__dirname, 'test-fixtures', 'normal-src', 'route.tsx');
  const option = {
    dirpath: path.join(__dirname, 'test-fixtures', 'normal-src/pages'),
    layoutDirPath: path.join(__dirname, 'test-fixtures', 'normal-src/layouts'),
    routeFilePath: outputFile,
    pathRewrite: {
      '^./': '@/',
    },
    hooks: {
      print: {
        parse: {
          before: printHooksParseBefore,
          after: printHooksParseAfter,
        },
        inject: {
          before: printHooksInjectBefore,
          after: printHooksInjectAfter,
        },
        write: {
          before: printHooksWriteBefore,
          after: printHooksWriteAfter,
        },
      },
    },
  };
  const sagaRoute = new SagaRoute(option);
  sagaRoute.routing();
  expect(printHooksParseBefore.mock.calls).toHaveLength(1);
  expect(printHooksParseAfter.mock.calls).toHaveLength(1);
  expect(printHooksInjectBefore.mock.calls).toHaveLength(1);
  expect(printHooksInjectAfter.mock.calls).toHaveLength(0);
  expect(printHooksWriteBefore.mock.calls).toHaveLength(0);
  expect(printHooksWriteAfter.mock.calls).toHaveLength(0);
});

test('test skip print.write', () => {
  const printHooksParseBefore = jest.fn();
  const printHooksParseAfter = jest.fn();
  const printHooksInjectBefore = jest.fn();
  const printHooksInjectAfter = jest.fn();
  const printHooksWriteBefore = jest.fn(() => null);
  const printHooksWriteAfter = jest.fn();
  const outputFile = path.join(__dirname, 'test-fixtures', 'normal-src', 'route.tsx');
  const option = {
    dirpath: path.join(__dirname, 'test-fixtures', 'normal-src/pages'),
    layoutDirPath: path.join(__dirname, 'test-fixtures', 'normal-src/layouts'),
    routeFilePath: outputFile,
    pathRewrite: {
      '^./': '@/',
    },
    hooks: {
      print: {
        parse: {
          before: printHooksParseBefore,
          after: printHooksParseAfter,
        },
        inject: {
          before: printHooksInjectBefore,
          after: printHooksInjectAfter,
        },
        write: {
          before: printHooksWriteBefore,
          after: printHooksWriteAfter,
        },
      },
    },
  };
  const sagaRoute = new SagaRoute(option);
  sagaRoute.routing();
  expect(printHooksParseBefore.mock.calls).toHaveLength(1);
  expect(printHooksParseAfter.mock.calls).toHaveLength(1);
  expect(printHooksInjectBefore.mock.calls).toHaveLength(1);
  expect(printHooksInjectAfter.mock.calls).toHaveLength(1);
  expect(printHooksWriteBefore.mock.calls).toHaveLength(1);
  expect(printHooksWriteAfter.mock.calls).toHaveLength(0);
});
