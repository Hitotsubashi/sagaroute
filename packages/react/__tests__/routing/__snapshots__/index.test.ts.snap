// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test routing in normal 1`] = `
"// @ts-nocheck
/* sagaroute-inject:imports */
/* injected by sagaroute: start */
import LayoutsIndex from "@/layouts/index";
import PagesIndex from "@/pages/index";
import Pages404 from "@/pages/404";
/* injected by sagaroute: end */

// sagaroute-inject: routes
const routes = [
  {
    "path": "/",
    "element": <LayoutsIndex/>,
    "children": [
      {
        "index": true,
        "element": <PagesIndex/>
      }
    ]
  },
  {
    "path": "*",
    "element": <Pages404/>
  }
];
export default routes;"
`;
